Auto Enable Captions based on user selected language

Optimus-NP opened the pull request:
Fixes: [#207], [#208]

### Description

This update enhances user experience by remembering the user's selected language and automatically enabling captions in that language when a TED video is loaded.

*Changes include:*

- Added utils.js to safely access `localStorage` (supports getItem, setItem, and removeItem).
- Updated app.js to write the selected language to `localStorage`.
- Modified article.js to:
  - Read the selected language on `window.onload`.
  - Automatically enable the caption track to match the stored language.


This ensures that captions in that language will be auto-enabled across video once a user selects a language.

### Testing Instructions

1. Convert TED talks to a ZIM file:

```bash
ted2zim --topics="wildlife" --debug --name="wildlife" --format=mp4 --title="Wildlife" --description="TED videos in Wildlife Category" --creator="TED" --publisher="openzim" --output="output" --keep --low-quality
```

2. Start the libkiwix server locally:

```bash
/usr/local/bin/kiwix-serve --port=8080 --library /tmp/zimlibrary/linux_library.xml --monitorLibrary
```

### Testing Instructions
1. Open the generated ZIM in Kiwix:
2. Select a language.
3. Refresh or load another video.
4. Confirm that captions are auto-enabled in the selected language.


https://github.com/user-attachments/assets/0f9fce1f-fe77-481e-af9d-e2a68718c907



### Note

Since `localStorage` might not always be available in all browser environments, it may be worth exploring server-side support for storing user preferences. As a future enhancement, I've proposed capturing **clickstream metrics** directly within the KiwixServer to better understand user behavior. See proposal: [kiwix/libkiwix#1191](https://github.com/kiwix/libkiwix/issues/1191)

Optimus-NP commented:
@kelson42 @benoit74 

Please review.

benoit74 commented on CHANGELOG.md:
```diff
@@ -14,6 +14,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add `--language-threshold` CLI argument for considering languages that appear in at least specified percentage of videos in `compute_zim_languages` (#212)
 - Add `--links` CLI argument for scraping videos from specific links (#237)
 - Circular back-to-top button and replaced text characters with icon buttons for better visual consistency (#112)
+- Auto Enable Captions based on user selected language (#246)
```
By convention we prefer to use issues descriptions and numbers ; it is acceptable to use PR when we do not have a linked issue or when PR is far wider than linked issue. Here you have to use issues (either in one or two changelog entries, does not matter much).

Optimus-NP replied:
Thank you. I have addressed this comment.

Optimus-NP replied:
Thanks for clarifying that storing preferences on the server side isn't the most efficient approach. I agree that using `localStorage` is the most suitable solution at this stage. I'll make sure to include this in the CHANGELOG.

benoit74 commented on src/ted2zim/templates/assets/article.js:
```diff
@@ -6,15 +6,35 @@ $.urlParam = function(name){
         $(".lang-default").css("display", "none");
         $(".lang-" + lang).css("display", "block");
+
+        // Initialize the video player and enable subtitles for the selected language
```
Langage is not selected but stored, please fix comment

Optimus-NP replied:
Thank you, Good catch.

benoit74 commented on src/ted2zim/templates/assets/article.js:
```diff
@@ -6,15 +6,35 @@ $.urlParam = function(name){
+            const player = this;
+            player.ready(() => {
+                const textTracks = player.textTracks();
+                // Iterate over all text tracks using a lambda expression
```
This comment does not help at all, please remove

Optimus-NP replied:
Thanks, Comment Addressed.

benoit74 commented on src/ted2zim/templates/assets/article.js:
```diff
@@ -6,15 +6,35 @@ $.urlParam = function(name){
+                Array.from(textTracks).forEach(t => {
+                    // If the track's language matches the selected language, show it
+                    // Otherwise, disable the subtitle track
+                    t.mode = (t.language === lang) ? 'showing' : 'disabled';
```
What does disabled mean? Is it still possible to select it manually?

What if two tracks matches the stored language, what happens?

What if audio is already in stored language ? I've specified in issue that we want to activate subtitle when storage language is set, stored language is available in a subtitle track **and** video audio is different than stored language.

Optimus-NP replied:
> **What does `disabled` mean? Is it still possible to select it manually?**  

`disabled` means the text track is inactive — it won’t display or update captions/subtitles. The `TextTrack#mode` property can have three values: `disabled`, `hidden`, or `showing`. When a track is `disabled`, it won’t render any captions.  [Ref1](https://docs.videojs.com/module-videojs-videojs.texttrack), [Ref2](https://html.spec.whatwg.org/multipage/media.html#text-track-disabled)



> **What if two tracks match the stored language, what happens?**  

Previously, the logic would pick the *last* matching track. I’ve now updated it to select the *first* matching track instead.  At this point, I haven’t handled cases where multiple tracks match more intelligently — open to suggestions here if there’s a cleaner way to approach this.



> **What if audio is already in stored language? I’ve specified in the issue that we want to activate subtitles when the stored language is set, a matching subtitle track is available, and the video audio is different from the stored language.**  

Good catch — I’ve handled this now. The audio language is passed from `scrapper.py` to `article.html`, and before enabling `showing` mode for the text track, I verify if the selected stored language matches the audio language. If they match, I skip enabling subtitles — ensuring they only activate when the video audio is different from the stored language.

Optimus-NP replied:
I’ve omitted explicitly setting the `mode` to `disabled`, as it is the `default` behavior.

benoit74 reviewed (changes requested):
Not yet tested, but I already have valuable feedback to share

benoit74 commented:
Regarding the fact that we might want to store things server-side, this is not possible. In general, ZIM users do not have access to a server. kiwix-serve is only one reader among many others, and we cannot / don't want to assume this is the main one. Not speaking about the issues linked to storing user stuff on a server and identifying user properly.

benoit74 commented:
One more remark: the key under which we store selected language must be more specific. In most readers, all ZIMs share the same local storage, so there is a significant risk of conflict. The fact that value is kept across ZIM updates is good. But I think we should use something a bit more specific like `ted2zim.selectedLanguage` to bring clarity when debugging local storage on a given device and avoid conflict. WDYT? This is something where we still have to learn / experiment tbh.

Optimus-NP commented:
> Not yet tested, but I already have valuable feedback to share

Thank you for the feedback @benoit74.

Optimus-NP commented:
> One more remark: the key under which we store selected language must be more specific. In most readers, all ZIMs share the same local storage, so there is a significant risk of conflict. The fact that value is kept across ZIM updates is good. But I think we should use something a bit more specific like `ted2zim.selectedLanguage` to bring clarity when debugging local storage on a given device and avoid conflict. WDYT? This is something where we still have to learn / experiment tbh.

That's an excellent suggestion, addressed the comment.

Optimus-NP commented:
> Regarding the fact that we might want to store things server-side, this is not possible. In general, ZIM users do not have access to a server. kiwix-serve is only one reader among many others, and we cannot / don't want to assume this is the main one. Not speaking about the issues linked to storing user stuff on a server and identifying user properly.

We are relying on `localStorage` to support this feature,  imagine if the client Interface (Browser leveraging kiwix-server, or Application Kiwix iOS, Application Kiwix Android, or Application Kiwix Desktop, etc) doesn't support `localStorage` then, we don't have any insight, that this feature isn't working. Hence, I planned to implement a backend service that keeps receiving metrics from the front end about the unsupported features.

benoit74 commented:
Most browsers support localStorage, and all our applications do support localStorage. The additional burden and consequences of having a server to store preferences is way beyond the benefit I see for now.

Optimus-NP commented on src/ted2zim/templates/assets/app.js:
```diff
@@ -1,5 +1,12 @@
-
 window.onload = function() {
+  // Check if a language is stored using the storage utility
+  let storedLanguage = storage.getItem(SELECTED_LANGUAGE_KEY);
```
Thank you for the feedback! I initially chose the variable name `storedLanguage` because the language was being read from `localStorage`, but you're right  `selectedLanguage` better reflects its purpose in the context of the UI. I'll update it accordingly.

benoit74 commented on src/ted2zim/templates/assets/article.js:
```diff
@@ -6,15 +6,42 @@ $.urlParam = function(name){
 };
 
+// Function to retrieve the stored language
+function getStoredLanguage() {
```
Please rename to `getSelectedLanguage` as well (sorry for not having found this specific issue earlier).

Optimus-NP replied:
More than happy to fix this. Highly Appreciated your feedback.

benoit74 reviewed (changes requested):
One last subtle change and we are good to go

benoit74 reviewed (approved):
LGTM, thank you !
